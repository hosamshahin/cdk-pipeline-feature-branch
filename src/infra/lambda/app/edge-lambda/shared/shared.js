"use strict";
// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchUrl = exports.ensureValidRedirectPath = exports.generateSecret = exports.RequiresConfirmationError = exports.timestampInSeconds = exports.sign = exports.urlSafe = exports.createErrorHtml = exports.httpPostToIDPWithRetry = exports.generateCookieHeaders = exports.extractAndParseCookies = exports.getCookieNames = exports.asCloudFrontHeaders = exports.getCompleteConfig = exports.getConfigWithHeaders = exports.getConfig = void 0;
const crypto_1 = require("crypto");
const fs_1 = require("fs");
const https_1 = require("https");
const https_2 = require("https");
const stream_1 = require("stream");
const util_1 = require("util");
const cookie_1 = require("cookie");
const template_html_1 = __importDefault(require("./error-page/template.html"));
const https_3 = require("./https");
function getDefaultCookieSettings(props) {
    // Defaults can be overridden by the user (CloudFormation Stack parameter) but should be solid enough for most purposes
    return {
        idToken: 'Path=/; Secure; SameSite=Lax',
        accessToken: 'Path=/; Secure; HttpOnly; SameSite=Lax',
        refreshToken: 'Path=/; Secure; HttpOnly; SameSite=Lax',
        nonce: 'Path=/; Secure; HttpOnly; SameSite=Lax',
    };
}
function isConfigWithHeaders(config) {
    return config.httpHeaders !== undefined;
}
function isCompleteConfig(config) {
    return true;
}
var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["none"] = 0] = "none";
    LogLevel[LogLevel["error"] = 10] = "error";
    LogLevel[LogLevel["warn"] = 20] = "warn";
    LogLevel[LogLevel["info"] = 30] = "info";
    LogLevel[LogLevel["debug"] = 40] = "debug";
})(LogLevel || (LogLevel = {}));
class Logger {
    constructor(logLevel) {
        this.logLevel = logLevel;
    }
    format(args, depth = 10) {
        return args.map((arg) => (0, util_1.formatWithOptions)({ depth }, arg)).join(' ');
    }
    info(...args) {
        if (this.logLevel >= LogLevel.info) {
            console.log(this.format(args));
        }
    }
    warn(...args) {
        if (this.logLevel >= LogLevel.warn) {
            console.warn(this.format(args));
        }
    }
    error(...args) {
        if (this.logLevel >= LogLevel.error) {
            console.error(this.format(args));
        }
    }
    debug(...args) {
        if (this.logLevel >= LogLevel.debug) {
            console.trace(this.format(args));
        }
    }
}
function getConfig() {
    const config = JSON.parse((0, fs_1.readFileSync)(`${__dirname}/configuration.json`).toString('utf8'));
    return {
        logger: new Logger(LogLevel[config.logLevel]),
        ...config,
    };
}
exports.getConfig = getConfig;
function getConfigWithHeaders() {
    const config = getConfig();
    if (!isConfigWithHeaders(config)) {
        throw new Error('Incomplete config in configuration.json');
    }
    return {
        cloudFrontHeaders: asCloudFrontHeaders(config.httpHeaders),
        ...config,
    };
}
exports.getConfigWithHeaders = getConfigWithHeaders;
function getCompleteConfig() {
    const config = getConfigWithHeaders();
    if (!isCompleteConfig(config)) {
        throw new Error('Incomplete config in configuration.json');
    }
    // Derive cookie settings by merging the defaults with the explicitly provided values
    const defaultCookieSettings = getDefaultCookieSettings({
        redirectPathAuthRefresh: config.redirectPathAuthRefresh,
    });
    const cookieSettings = config.cookieSettings
        ? Object.fromEntries(Object.entries({
            ...defaultCookieSettings,
            ...config.cookieSettings,
        }).map(([k, v]) => [
            k,
            v || defaultCookieSettings[k],
        ]))
        : defaultCookieSettings;
    // Defaults for nonce and PKCE
    const defaults = {
        secretAllowedCharacters: 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~',
        pkceLength: 43,
        nonceLength: 16,
        nonceMaxAge: (cookieSettings?.nonce &&
            parseInt((0, cookie_1.parse)(cookieSettings.nonce.toLowerCase())['max-age'])) ||
            60 * 60 * 24,
    };
    return {
        ...defaults,
        ...config,
        cookieSettings,
    };
}
exports.getCompleteConfig = getCompleteConfig;
function extractCookiesFromHeaders(headers) {
    // Cookies are present in the HTTP header "Cookie" that may be present multiple times.
    // This utility function parses occurrences  of that header and splits out all the cookies and their values
    // A simple object is returned that allows easy access by cookie name: e.g. cookies["nonce"]
    if (!headers.cookie) {
        return {};
    }
    const cookies = headers.cookie.reduce((reduced, header) => Object.assign(reduced, (0, cookie_1.parse)(header.value)), {});
    return cookies;
}
function asCloudFrontHeaders(headers) {
    if (!headers)
        return {};
    // Turn a regular key-value object into the explicit format expected by CloudFront
    return Object.entries(headers).reduce((reduced, [key, value]) => Object.assign(reduced, {
        [key.toLowerCase()]: [
            {
                key,
                value,
            },
        ],
    }), {});
}
exports.asCloudFrontHeaders = asCloudFrontHeaders;
function getCookieNames(clientId, idTokenCookieName) {
    const keyPrefix = `idp.${clientId}`;
    return {
        lastUserKey: `${keyPrefix}.LastAuthUser`,
        scopeKey: `${keyPrefix}.tokenScopesString`,
        idTokenKey: `${keyPrefix}.idToken`,
        idToken2Key: idTokenCookieName,
        accessTokenKey: `${keyPrefix}.accessToken`,
        refreshTokenKey: `${keyPrefix}.refreshToken`,
    };
}
exports.getCookieNames = getCookieNames;
function extractAndParseCookies(headers, clientId, idTokenCookieName) {
    const cookies = extractCookiesFromHeaders(headers);
    if (!cookies) {
        return {};
    }
    let cookieNames;
    cookieNames = getCookieNames(clientId, idTokenCookieName);
    return {
        tokenUserName: cookies[cookieNames.lastUserKey],
        idToken: cookies[cookieNames.idTokenKey],
        idToken2: cookies[cookieNames.idToken2Key],
        accessToken: cookies[cookieNames.accessTokenKey],
        refreshToken: cookies[cookieNames.refreshTokenKey],
        scopes: cookies[cookieNames.scopeKey],
        nonce: cookies['spa-auth-edge-nonce'],
        nonceHmac: cookies['spa-auth-edge-nonce-hmac'],
        pkce: cookies['spa-auth-edge-pkce'],
    };
}
exports.extractAndParseCookies = extractAndParseCookies;
exports.generateCookieHeaders = {
    signIn: (param) => _generateCookieHeaders({ ...param, event: 'signIn' }),
    refresh: (param) => _generateCookieHeaders({ ...param, event: 'refresh' }),
    signOut: (param) => _generateCookieHeaders({ ...param, event: 'signOut' }),
};
function _generateCookieHeaders(param) {
    /**
     * Generate cookie headers for the following scenario's:
     *  - signIn: called from Parse Auth lambda, when receiving fresh JWTs from IDP
     *  - sign out: called from Sign Out Lambda, when the user visits the sign out URL
     *  - refresh: called from Refresh Auth lambda, when receiving fresh ID and Access JWTs from IDP
     *
     *   Note that there are other places besides this helper function where cookies can be set (search codebase for "set-cookie")
     */
    const cookies = {};
    let cookieNames;
    cookieNames = getCookieNames(param.clientId, param.idTokenCookieName);
    // Construct object with the cookies
    Object.assign(cookies, {
        [cookieNames.scopeKey]: `${param.oauthScopes.join(' ')}; ${param.cookieSettings.accessToken}`,
    });
    // Set JWTs in the cookies
    cookies[cookieNames.accessTokenKey] = `${param.tokens.access}; ${param.cookieSettings.accessToken}`;
    if (param.tokens.id) {
        cookies[cookieNames.idTokenKey] = `${param.tokens.id}; ${param.cookieSettings.idToken}`;
        cookies[cookieNames.idToken2Key] = `${param.tokens.id}; ${param.cookieSettings.idToken}`;
    }
    if (param.tokens.refresh) {
        cookies[cookieNames.refreshTokenKey] = `${param.tokens.refresh}; ${param.cookieSettings.refreshToken}`;
    }
    if (param.event === 'signOut') {
        // Expire all cookies
        Object.keys(cookies).forEach((key) => (cookies[key] = expireCookie(cookies[key])));
    }
    // Always expire nonce, nonceHmac and pkce - this is valid in all scenario's:
    // * event === 'newTokens' --> you just signed in and used your nonce and pkce successfully, don't need them no more
    // * event === 'refreshFailed' --> you are signed in already, why do you still have a nonce?
    // * event === 'signOut' --> clear ALL cookies anyway
    [
        'spa-auth-edge-nonce',
        'spa-auth-edge-nonce-hmac',
        'spa-auth-edge-pkce',
    ].forEach((key) => {
        cookies[key] = expireCookie(`;${param.cookieSettings.nonce}`);
    });
    // Return cookie object in format of CloudFront headers
    return Object.entries({
        ...cookies,
    }).map(([k, v]) => ({ key: 'set-cookie', value: `${k}=${v}` }));
}
function expireCookie(cookie = '') {
    const cookieParts = cookie
        .split(';')
        .map((part) => part.trim())
        .filter((part) => !part.toLowerCase().startsWith('max-age'))
        .filter((part) => !part.toLowerCase().startsWith('expires'));
    const expires = `Expires=${new Date(0).toUTCString()}`;
    const [, ...settings] = cookieParts; // first part is the cookie value, which we'll clear
    return ['', ...settings, expires].join('; ');
}
function decodeToken(jwt) {
    const tokenBody = jwt.split('.')[1];
    const decodableTokenBody = tokenBody.replace(/-/g, '+').replace(/_/g, '/');
    return JSON.parse(Buffer.from(decodableTokenBody, 'base64').toString());
}
const AGENT = new https_1.Agent({ keepAlive: true });
async function httpPostToIDPWithRetry(url, data, options, logger, method) {
    let attempts = 0;
    while (true) {
        ++attempts;
        try {
            return await (0, https_3.fetch)(url, data, {
                agent: AGENT,
                ...options,
                method,
            }).then((res) => {
                if (res.status !== 200) {
                    throw new Error(`Status is ${res.status}, expected 200`);
                }
                if (!res.headers['content-type']?.startsWith('application/json')) {
                    throw new Error(`Content-Type is ${res.headers['content-type']}, expected application/json`);
                }
                return {
                    ...res,
                    data: JSON.parse(res.data.toString()),
                };
            });
        }
        catch (err) {
            logger.debug(`HTTP ${method} to ${url} failed (attempt ${attempts}):`);
            logger.debug(err);
            if (attempts >= 5) {
                // Try 5 times at most
                logger.error(`No success after ${attempts} attempts, seizing further attempts`);
                throw err;
            }
            if (attempts >= 2) {
                // After attempting twice immediately, do some exponential backoff with jitter
                logger.debug(`Doing exponential backoff with jitter, before attempting HTTP ${method} again ...`);
                await new Promise((resolve) => setTimeout(resolve, 25 * (Math.pow(2, attempts) + Math.random() * attempts)));
                logger.debug(`Done waiting, will try HTTP ${method} again now`);
            }
        }
    }
}
exports.httpPostToIDPWithRetry = httpPostToIDPWithRetry;
function createErrorHtml(props) {
    const params = { ...props, region: process.env.AWS_REGION };
    return template_html_1.default.replace(/\${([^}]*)}/g, (_, v) => escapeHtml(params[v]) ?? '');
}
exports.createErrorHtml = createErrorHtml;
function escapeHtml(unsafe) {
    if (typeof unsafe !== 'string') {
        return undefined;
    }
    return unsafe
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#039;');
}
exports.urlSafe = {
    /*
          Functions to translate base64-encoded strings, so they can be used:
          - in URL's without needing additional encoding
          - in OAuth2 PKCE verifier
          - in cookies (to be on the safe side, as = + / are in fact valid characters in cookies)
  
          stringify:
              use this on a base64-encoded string to translate = + / into replacement characters
  
          parse:
              use this on a string that was previously urlSafe.stringify'ed to return it to
              its prior pure-base64 form. Note that trailing = are not added, but NodeJS does not care
      */
    stringify: (b64encodedString) => b64encodedString.replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_'),
    parse: (b64encodedString) => b64encodedString.replace(/-/g, '+').replace(/_/g, '/'),
};
function sign(stringToSign, secret, signatureLength) {
    const digest = (0, crypto_1.createHmac)('sha256', secret)
        .update(stringToSign)
        .digest('base64')
        .slice(0, signatureLength);
    const signature = exports.urlSafe.stringify(digest);
    return signature;
}
exports.sign = sign;
function timestampInSeconds() {
    return (Date.now() / 1000) | 0;
}
exports.timestampInSeconds = timestampInSeconds;
class RequiresConfirmationError extends Error {
}
exports.RequiresConfirmationError = RequiresConfirmationError;
function generateSecret(allowedCharacters, secretLength) {
    return [...new Array(secretLength)]
        .map(() => allowedCharacters[(0, crypto_1.randomInt)(0, allowedCharacters.length)])
        .join('');
}
exports.generateSecret = generateSecret;
function ensureValidRedirectPath(path) {
    if (typeof path !== 'string')
        return '/';
    return path.startsWith('/') ? path : `/${path}`;
}
exports.ensureValidRedirectPath = ensureValidRedirectPath;
async function fetchUrl(uri) {
    return new Promise((resolve, reject) => {
        const req = (0, https_2.request)(uri, (res) => (0, stream_1.pipeline)([res, collectBuffer(resolve)], done));
        function done(error) {
            if (!error)
                return;
            req.destroy(error);
            reject(error);
        }
        req.on('error', done);
        req.end();
    });
}
exports.fetchUrl = fetchUrl;
const collectBuffer = (callback) => {
    const chunks = [];
    return new stream_1.Writable({
        write: (chunk, _encoding, done) => {
            try {
                chunks.push(chunk);
                done();
            }
            catch (err) {
                done(err);
            }
        },
        final: (done) => {
            try {
                callback(Buffer.concat(chunks));
                done();
            }
            catch (err) {
                done(err);
            }
        },
    });
};
