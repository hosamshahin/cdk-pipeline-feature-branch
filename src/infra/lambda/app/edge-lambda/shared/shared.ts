// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import { createHmac, randomInt } from 'crypto';
import { readFileSync } from 'fs';
import { Agent, RequestOptions } from 'https';
import { request } from 'https';
import { Writable, pipeline } from 'stream';
import { formatWithOptions } from 'util';
import { CloudFrontHeaders } from 'aws-lambda';
import { parse } from 'cookie';
import html from './error-page/template.html';
import { fetch } from './https';

export interface CookieSettings {
  idToken: string;
  accessToken: string;
  refreshToken: string;
  nonce: string;
  [key: string]: string;
}

function getDefaultCookieSettings(props: {
  redirectPathAuthRefresh: string;
}): CookieSettings {
  // Defaults can be overridden by the user (CloudFormation Stack parameter) but should be solid enough for most purposes
  return {
    idToken: 'Path=/; Secure; SameSite=Lax',
    accessToken: 'Path=/; Secure; HttpOnly; SameSite=Lax',
    refreshToken: 'Path=/; Secure; HttpOnly; SameSite=Lax',
    nonce: 'Path=/; Secure; HttpOnly; SameSite=Lax',
  };
}

export interface HttpHeaders {
  [key: string]: string;
}

interface ConfigFromDisk {
  logLevel: keyof typeof LogLevel;
}

interface ConfigFromDiskWithHeaders extends ConfigFromDisk {
  httpHeaders: HttpHeaders;
}

interface ConfigFromDiskComplete extends ConfigFromDiskWithHeaders {
  userPoolArn: string;
  clientId: string;
  oauthScopes: string[];
  authEndpoint: string;
  accessTokenEndpoint: string;
  introspectEndpoint: string;
  pingEndSessionEndpoint: string;
  redirectPathSignIn: string;
  redirectPathSignOut: string;
  signOutUrl: string;
  redirectPathAuthRefresh: string;
  cookieSettings: CookieSettings;
  clientSecret: string;
  nonceSigningSecret: string;
  secretAllowedCharacters?: string;
  pkceLength?: number;
  nonceLength?: number;
  nonceMaxAge?: number;
  idTokenCookieName: string;
  staticContentPathPattern: string,
  staticContentRootObject: string
}

function isConfigWithHeaders(config: any): config is ConfigFromDiskComplete {
  return config.httpHeaders !== undefined;
}

function isCompleteConfig(config: any): config is ConfigFromDiskComplete {
  return true;
}

enum LogLevel {
  'none' = 0,
  'error' = 10,
  'warn' = 20,
  'info' = 30,
  'debug' = 40,
}

class Logger {
  constructor(private logLevel: LogLevel) { }

  private format(args: unknown[], depth = 10) {
    return args.map((arg) => formatWithOptions({ depth }, arg)).join(' ');
  }

  public info(...args: unknown[]) {
    if (this.logLevel >= LogLevel.info) {
      console.log(this.format(args));
    }
  }
  public warn(...args: unknown[]) {
    if (this.logLevel >= LogLevel.warn) {
      console.warn(this.format(args));
    }
  }
  public error(...args: unknown[]) {
    if (this.logLevel >= LogLevel.error) {
      console.error(this.format(args));
    }
  }
  public debug(...args: unknown[]) {
    if (this.logLevel >= LogLevel.debug) {
      console.trace(this.format(args));
    }
  }
}

export interface Config extends ConfigFromDisk {
  logger: Logger;
}

export interface ConfigWithHeaders extends Config, ConfigFromDiskWithHeaders {
  cloudFrontHeaders: CloudFrontHeaders;
}

export interface CompleteConfig
  extends ConfigWithHeaders,
  ConfigFromDiskComplete {
  cloudFrontHeaders: CloudFrontHeaders;
  secretAllowedCharacters: string;
  pkceLength: number;
  nonceLength: number;
  nonceMaxAge: number;
}

export function getConfig(): Config {
  const config = JSON.parse(
    readFileSync(`${__dirname}/configuration.json`).toString('utf8'),
  ) as ConfigFromDisk;
  return {
    logger: new Logger(LogLevel[config.logLevel]),
    ...config,
  };
}

export function getConfigWithHeaders(): ConfigWithHeaders {
  const config = getConfig();

  if (!isConfigWithHeaders(config)) {
    throw new Error('Incomplete config in configuration.json');
  }

  return {
    cloudFrontHeaders: asCloudFrontHeaders(config.httpHeaders),
    ...config,
  };
}

export function getCompleteConfig(): CompleteConfig {
  const config = getConfigWithHeaders();

  if (!isCompleteConfig(config)) {
    throw new Error('Incomplete config in configuration.json');
  }

  // Derive cookie settings by merging the defaults with the explicitly provided values
  const defaultCookieSettings = getDefaultCookieSettings({
    redirectPathAuthRefresh: config.redirectPathAuthRefresh,
  });
  const cookieSettings = config.cookieSettings
    ? (Object.fromEntries(
      Object.entries({
        ...defaultCookieSettings,
        ...config.cookieSettings,
      }).map(([k, v]) => [
        k,
        v || defaultCookieSettings[k as keyof CookieSettings],
      ]),
    ) as CookieSettings)
    : defaultCookieSettings;

  // Defaults for nonce and PKCE
  const defaults = {
    secretAllowedCharacters:
      'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~',
    pkceLength: 43, // Should be between 43 and 128 - per spec
    nonceLength: 16,
    nonceMaxAge:
      (cookieSettings?.nonce &&
        parseInt(parse(cookieSettings.nonce.toLowerCase())['max-age'])) ||
      60 * 60 * 24,
  };

  return {
    ...defaults,
    ...config,
    cookieSettings,
  };
}


type Cookies = { [key: string]: string };

function extractCookiesFromHeaders(headers: CloudFrontHeaders) {
  // Cookies are present in the HTTP header "Cookie" that may be present multiple times.
  // This utility function parses occurrences  of that header and splits out all the cookies and their values
  // A simple object is returned that allows easy access by cookie name: e.g. cookies["nonce"]
  if (!headers.cookie) {
    return {};
  }
  const cookies = headers.cookie.reduce(
    (reduced, header) => Object.assign(reduced, parse(header.value)),
    {} as Cookies,
  );

  return cookies;
}

export function asCloudFrontHeaders(headers: HttpHeaders): CloudFrontHeaders {
  if (!headers) return {};
  // Turn a regular key-value object into the explicit format expected by CloudFront
  return Object.entries(headers).reduce(
    (reduced, [key, value]) =>
      Object.assign(reduced, {
        [key.toLowerCase()]: [
          {
            key,
            value,
          },
        ],
      }),
    {} as CloudFrontHeaders,
  );
}

export function getCookieNames(clientId: string, idTokenCookieName: string) {
  const keyPrefix = `idp.${clientId}`;
  return {
    lastUserKey: `${keyPrefix}.LastAuthUser`,
    scopeKey: `${keyPrefix}.tokenScopesString`,
    idTokenKey: `${keyPrefix}.idToken`,
    idToken2Key: idTokenCookieName,
    accessTokenKey: `${keyPrefix}.accessToken`,
    refreshTokenKey: `${keyPrefix}.refreshToken`,
  };
}

export function extractAndParseCookies(
  headers: CloudFrontHeaders,
  clientId: string,
  idTokenCookieName: string,
) {
  const cookies = extractCookiesFromHeaders(headers);
  if (!cookies) {
    return {};
  }

  let cookieNames: { [name: string]: string };

  cookieNames = getCookieNames(clientId, idTokenCookieName);

  return {
    tokenUserName: cookies[cookieNames.lastUserKey],
    idToken: cookies[cookieNames.idTokenKey],
    idToken2: cookies[cookieNames.idToken2Key],
    accessToken: cookies[cookieNames.accessTokenKey],
    refreshToken: cookies[cookieNames.refreshTokenKey],
    scopes: cookies[cookieNames.scopeKey],
    nonce: cookies['spa-auth-edge-nonce'],
    nonceHmac: cookies['spa-auth-edge-nonce-hmac'],
    pkce: cookies['spa-auth-edge-pkce'],
  };
}

interface GenerateCookieHeadersParam {
  clientId: string;
  oauthScopes: string[];
  cookieSettings: CookieSettings;
  idTokenCookieName: string;
  tokens: {
    id?: string;
    access: string;
    refresh?: string;
  };
}

export const generateCookieHeaders = {
  signIn: (
    param: GenerateCookieHeadersParam & {
      tokens: { id: string; access: string; refresh: string };
    },
  ) => _generateCookieHeaders({ ...param, event: 'signIn' }),
  refresh: (
    param: GenerateCookieHeadersParam & {
      tokens: { access: string };
    },
  ) => _generateCookieHeaders({ ...param, event: 'refresh' }),
  signOut: (param: GenerateCookieHeadersParam) =>
    _generateCookieHeaders({ ...param, event: 'signOut' }),
};

function _generateCookieHeaders(
  param: GenerateCookieHeadersParam & {
    event: 'signIn' | 'signOut' | 'refresh';
  },
) {
  /**
   * Generate cookie headers for the following scenario's:
   *  - signIn: called from Parse Auth lambda, when receiving fresh JWTs from IDP
   *  - sign out: called from Sign Out Lambda, when the user visits the sign out URL
   *  - refresh: called from Refresh Auth lambda, when receiving fresh ID and Access JWTs from IDP
   *
   *   Note that there are other places besides this helper function where cookies can be set (search codebase for "set-cookie")
   */

  const cookies: Cookies = {};
  let cookieNames: ReturnType<typeof getCookieNames>;

  cookieNames = getCookieNames(param.clientId, param.idTokenCookieName);

  // Construct object with the cookies
  Object.assign(cookies, {
    [cookieNames.scopeKey]: `${param.oauthScopes.join(' ')}; ${param.cookieSettings.accessToken}`,
  });

  // Set JWTs in the cookies
  cookies[cookieNames.accessTokenKey] = `${param.tokens.access}; ${param.cookieSettings.accessToken}`;

  if (param.tokens.id) {
    cookies[cookieNames.idTokenKey] = `${param.tokens.id}; ${param.cookieSettings.idToken}`;
    cookies[cookieNames.idToken2Key] = `${param.tokens.id}; ${param.cookieSettings.idToken}`;
  }

  if (param.tokens.refresh) {
    cookies[cookieNames.refreshTokenKey] = `${param.tokens.refresh}; ${param.cookieSettings.refreshToken}`;
  }

  if (param.event === 'signOut') {
    // Expire all cookies
    Object.keys(cookies).forEach(
      (key) => (cookies[key] = expireCookie(cookies[key])),
    );
  }

  // Always expire nonce, nonceHmac and pkce - this is valid in all scenario's:
  // * event === 'newTokens' --> you just signed in and used your nonce and pkce successfully, don't need them no more
  // * event === 'refreshFailed' --> you are signed in already, why do you still have a nonce?
  // * event === 'signOut' --> clear ALL cookies anyway
  [
    'spa-auth-edge-nonce',
    'spa-auth-edge-nonce-hmac',
    'spa-auth-edge-pkce',
  ].forEach((key) => {
    cookies[key] = expireCookie(`;${param.cookieSettings.nonce}`);
  });

  // Return cookie object in format of CloudFront headers
  return Object.entries({
    ...cookies,
  }).map(([k, v]) => ({ key: 'set-cookie', value: `${k}=${v}` }));
}

function expireCookie(cookie: string = '') {
  const cookieParts = cookie
    .split(';')
    .map((part) => part.trim())
    .filter((part) => !part.toLowerCase().startsWith('max-age'))
    .filter((part) => !part.toLowerCase().startsWith('expires'));
  const expires = `Expires=${new Date(0).toUTCString()}`;
  const [, ...settings] = cookieParts; // first part is the cookie value, which we'll clear
  return ['', ...settings, expires].join('; ');
}

function decodeToken(jwt: string) {
  const tokenBody = jwt.split('.')[1];
  const decodableTokenBody = tokenBody.replace(/-/g, '+').replace(/_/g, '/');
  return JSON.parse(Buffer.from(decodableTokenBody, 'base64').toString());
}

const AGENT = new Agent({ keepAlive: true });

export async function httpPostToIDPWithRetry(
  url: string,
  data: Buffer,
  options: RequestOptions,
  logger: Logger,
  method: 'POST' | 'GET',
) {
  let attempts = 0;
  while (true) {
    ++attempts;
    try {
      return await fetch(url, data, {
        agent: AGENT,
        ...options,
        method,
      }).then((res) => {
        if (res.status !== 200) {
          throw new Error(`Status is ${res.status}, expected 200`);
        }
        if (!res.headers['content-type']?.startsWith('application/json')) {
          throw new Error(
            `Content-Type is ${res.headers['content-type']}, expected application/json`,
          );
        }
        return {
          ...res,
          data: JSON.parse(res.data.toString()),
        };
      });
    } catch (err) {
      logger.debug(`HTTP ${method} to ${url} failed (attempt ${attempts}):`);
      logger.debug(err);
      if (attempts >= 5) {
        // Try 5 times at most
        logger.error(
          `No success after ${attempts} attempts, seizing further attempts`,
        );
        throw err;
      }
      if (attempts >= 2) {
        // After attempting twice immediately, do some exponential backoff with jitter
        logger.debug(
          `Doing exponential backoff with jitter, before attempting HTTP ${method} again ...`,
        );
        await new Promise((resolve) =>
          setTimeout(
            resolve,
            25 * (Math.pow(2, attempts) + Math.random() * attempts),
          ),
        );
        logger.debug(`Done waiting, will try HTTP ${method} again now`);
      }
    }
  }
}

export function createErrorHtml(props: {
  title: string;
  message: string;
  expandText?: string;
  details?: string;
  linkUri: string;
  linkText: string;
}) {
  const params = { ...props, region: process.env.AWS_REGION };
  return html.replace(
    /\${([^}]*)}/g,
    (_: any, v: keyof typeof params) => escapeHtml(params[v]) ?? '',
  );
}

function escapeHtml(unsafe: unknown) {
  if (typeof unsafe !== 'string') {
    return undefined;
  }
  return unsafe
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#039;');
}

export const urlSafe = {
  /*
        Functions to translate base64-encoded strings, so they can be used:
        - in URL's without needing additional encoding
        - in OAuth2 PKCE verifier
        - in cookies (to be on the safe side, as = + / are in fact valid characters in cookies)

        stringify:
            use this on a base64-encoded string to translate = + / into replacement characters

        parse:
            use this on a string that was previously urlSafe.stringify'ed to return it to
            its prior pure-base64 form. Note that trailing = are not added, but NodeJS does not care
    */
  stringify: (b64encodedString: string) =>
    b64encodedString.replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_'),
  parse: (b64encodedString: string) =>
    b64encodedString.replace(/-/g, '+').replace(/_/g, '/'),
};

export function sign(
  stringToSign: string,
  secret: string,
  signatureLength: number,
) {
  const digest = createHmac('sha256', secret)
    .update(stringToSign)
    .digest('base64')
    .slice(0, signatureLength);
  const signature = urlSafe.stringify(digest);
  return signature;
}

export function timestampInSeconds() {
  return (Date.now() / 1000) | 0;
}

export class RequiresConfirmationError extends Error { }

export function generateSecret(
  allowedCharacters: string,
  secretLength: number,
) {
  return [...new Array(secretLength)]
    .map(() => allowedCharacters[randomInt(0, allowedCharacters.length)])
    .join('');
}

export function ensureValidRedirectPath(path: unknown) {
  if (typeof path !== 'string') return '/';
  return path.startsWith('/') ? path : `/${path}`;
}

export async function fetchUrl(uri: string) {
  return new Promise<Buffer>((resolve, reject) => {
    const req = request(uri, (res) =>
      pipeline([res, collectBuffer(resolve)], done),
    );

    function done(error?: Error | null) {
      if (!error) return;
      req.destroy(error);
      reject(error);
    }

    req.on('error', done);

    req.end();
  });
}

const collectBuffer = (callback: (collectedBuffer: Buffer) => void) => {
  const chunks = [] as Buffer[];
  return new Writable({
    write: (chunk, _encoding, done) => {
      try {
        chunks.push(chunk);
        done();
      } catch (err) {
        done(err as Error);
      }
    },
    final: (done) => {
      try {
        callback(Buffer.concat(chunks));
        done();
      } catch (err) {
        done(err as Error);
      }
    },
  });
};