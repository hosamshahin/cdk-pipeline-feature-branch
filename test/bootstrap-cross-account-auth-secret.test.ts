import * as cdk from 'aws-cdk-lib';
import { Annotations, Match, Template } from 'aws-cdk-lib/assertions';
import { AwsSolutionsChecks, NagSuppressions } from 'cdk-nag';
import * as yaml from 'yaml';
import { AuthSecret } from '../src/infra/shared/bootstrap-cross-account-auth-secret';

test('AuthSecret cdk-nag AwsSolutions Pack', () => {

  const app = new cdk.App();
  const env = {
    account: process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEFAULT_REGION,
  };

  const stack = new AuthSecret(app, 'AuthSecret', { env });

  cdk.Aspects.of(stack).add(new AwsSolutionsChecks());

  NagSuppressions.addStackSuppressions(stack, [
    { id: 'AwsSolutions-SMG4', reason: 'TBD reason of 10 characters or more' },
  ]);

  const warnings = Annotations.fromStack(stack).findWarning(
    '*',
    Match.stringLikeRegexp('AwsSolutions-.*'),
  );

  // console.dir(warnings)
  expect(warnings).toHaveLength(0);

  const errors = Annotations.fromStack(stack).findError(
    '*',
    Match.stringLikeRegexp('AwsSolutions-.*'),
  );

  // console.dir(errors)
  expect(errors).toHaveLength(0);

  const template = Template.fromStack(stack);
  expect(yaml.stringify(template)).toMatchSnapshot();

});
