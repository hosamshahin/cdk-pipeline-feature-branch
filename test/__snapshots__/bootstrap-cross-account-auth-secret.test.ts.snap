// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthSecret cdk-nag AwsSolutions Pack 1`] = `
"Metadata:
  cdk_nag:
    rules_to_suppress:
      - reason: TBD reason of 10 characters or more
        id: AwsSolutions-SMG4
Resources:
  AuthSecretC13F30CE:
    Type: AWS::SecretsManager::Secret
    Properties:
      GenerateSecretString:
        ExcludePunctuation: true
        GenerateStringKey: clientSecret
        IncludeSpace: false
        SecretStringTemplate: '{"clientId":""}'
    UpdateReplacePolicy: Delete
    DeletionPolicy: Delete
Outputs:
  AuthSecretOutput:
    Value:
      Ref: AuthSecretC13F30CE
    Export:
      Name: AuthSecretOutput
Parameters:
  BootstrapVersion:
    Type: AWS::SSM::Parameter::Value<String>
    Default: /cdk-bootstrap/hnb659fds/version
    Description: Version of the CDK Bootstrap resources in this environment,
      automatically retrieved from SSM Parameter Store. [cdk:skip]
Rules:
  CheckBootstrapVersion:
    Assertions:
      - Assert:
          Fn::Not:
            - Fn::Contains:
                - - "1"
                  - "2"
                  - "3"
                  - "4"
                  - "5"
                - Ref: BootstrapVersion
        AssertDescription: CDK bootstrap stack version 6 required. Please run 'cdk
          bootstrap' with a recent version of the CDK CLI.
"
`;
